Add Multi-User Parallel Development Support to AI-DLC

jschwellach opened the pull request:
## Overview

This PR introduces comprehensive multi-user parallel development capabilities to AI-DLC, enabling teams to collaborate efficiently on software projects with automatic workflow management and conflict prevention.

## Key Features

### 🔀 **Automatic Branch Management**
- Each developer works in isolated branches ()
- Automatic branch creation and switching
- Clean separation of work streams

### 📝 **User-Scoped Audit Files**
- Prevents audit file conflicts between team members
- Each user maintains their own audit trail
- Consolidated project-wide visibility

### 🎯 **Unit Assignment Tracking**
- Clear visibility of who's working on what
- Prevents duplicate work assignments
- Real-time status tracking

### 🔄 **Dependency-Aware Consolidation**
- Automatic merging in correct dependency order
- Intelligent conflict resolution
- Maintains project integrity

### 📊 **Consolidated Project Views**
- Project-wide visibility with automatic indexing
- Cross-unit dependency tracking
- Unified progress monitoring

## Workflow Integration

The multi-user capabilities seamlessly integrate with existing AI-DLC phases:

1. **Inception Phase**: Lead developer completes requirements and creates units
2. **Team Collaboration**: Multiple developers claim and work on units in parallel
3. **Construction Phase**: Each developer implements their assigned units
4. **Consolidation**: Automatic merging with dependency resolution

## Usage Examples

```bash
# Lead completes inception and enables team collaboration
kiro-cli chat "Using AI-DLC, create microservices architecture"
# → Choose "Enable team collaboration"

# Team members claim units
kiro-cli chat "claim unit user-service"
kiro-cli chat "claim unit payment-service"

# Each developer works in their isolated branch
# unit/user-service-alice, unit/payment-service-bob

# Integrator consolidates when ready
kiro-cli chat "consolidate units"
```

## Technical Implementation

### New Files Added
- `multi-user-parallel.md`: Core multi-user workflow logic
- `units-generation.md`: Enhanced unit creation with team support
- `AGENTS.md.example`: Configuration template for team setups

### Updated Components
- Enhanced AI-DLC rules with multi-user decision points
- Automatic branch management workflows
- User-scoped file generation and audit trails
- Dependency-aware consolidation logic

## Configuration

Teams can customize behavior via `AGENTS.md`:
- Generated code folder structure per unit
- Audit file locations and naming
- Branch naming conventions
- Custom paths for project organization

## Benefits

- **Parallel Development**: Multiple developers work simultaneously without conflicts
- **Reduced Bottlenecks**: No waiting for sequential completion
- **Quality Assurance**: Isolated work streams with proper integration
- **Scalability**: Supports teams of any size
- **Flexibility**: Configurable to match team workflows

## Compatibility

- Fully backward compatible with single-developer workflows
- Works with existing AI-DLC projects
- No breaking changes to current functionality
- Optional team features activated only when needed

## Testing

Tested with:
- Single developer workflows (existing functionality)
- Multi-developer parallel development
- Complex dependency scenarios
- Branch conflict resolution
- Audit file isolation

## Migration Notes

- Existing projects continue to work unchanged
- New projects can opt into multi-user features
- Configuration is optional and project-specific
- No impact on current AI-DLC installations

This enhancement transforms AI-DLC from a single-developer tool into a comprehensive team collaboration platform while maintaining its intelligent, adaptive workflow characteristics.

raj-jain-aws commented on AGENTS.md.example:
AGENTS.md will surely give us tool independence, but can we think about moving the entire AI-DLC steering as part of AGENTS.md as a separate feature? I think we also discussed there are other issues around AGENTS.md that if the user already has an AGENTS.md then how this will be merged into their existing file,

raj-jain-aws commented on aidlc-rules/aws-aidlc-rule-details/inception/workflow-planning.md:
```diff
@@ -202,23 +202,20 @@ Create Mermaid flowchart showing:
-- Conditional SKIP: `fill:#BDBDBD,stroke:#424242,stroke-width:2px,stroke-dasharray: 5 5,color:#000` (Material Gray with black text)
-- Start/End: `fill:#CE93D8,stroke:#6A1B9A,stroke-width:3px,color:#000` (Material Purple with black text)
-- Phase containers: Use lighter Material colors (INCEPTION: #BBDEFB, CONSTRUCTION: #C8E6C9, OPERATIONS: #FFF59D)
+- Completed/Always execute: `fill:#90EE90,stroke:#2d5016,stroke-width:3px`
```
Suggest we propose only one logical change per PR. Mermaid styling change can be proposed as part of another PR.

raj-jain-aws commented on aidlc-rules/aws-aidlc-rule-details/inception/units-generation.md:
```diff
@@ -152,6 +152,86 @@ If the analysis in step 7 reveals ANY ambiguous answers, you MUST:
 - Update the "Current Status" section
 - Prepare for transition to CONSTRUCTION PHASE
 
+## Step 20: Setup Multi-User Parallel Development
```
I tested this PR and did not see this step kick in. I've ensured that I've the files from this this PR. I see the following where the workflow is entering Unit 1 and is not offering me multi-unit questions.

```
Perfect! The INCEPTION PHASE is now complete. We've successfully:

✅ Workspace Detection - Identified greenfield project ✅ Requirements Analysis - Defined 32 functional requirements and 8 NFR categories
✅ Workflow Planning - Created execution plan with 10 stages ✅ Application Design - Designed 20 components across 5 architectural layers ✅ Units Generation - Decomposed into 5 units with clear dependencies

🟢 CONSTRUCTION PHASE Begins
We're now entering the CONSTRUCTION PHASE where we'll design and implement each unit. The workflow will proceed through each unit in dependency order:

Development Sequence:

Unit 1: Shared Components (foundation - starting now)
Unit 2: User Management (authentication)
Units 3 & 4: Document Processing + Requirements Generation (parallel)
Unit 5: Frontend Portal (integration)
Build and Test (final validation)
For each unit, we'll execute:

Functional Design (detailed business logic)
NFR Requirements (performance, security, scalability)
NFR Design (patterns and implementations)
Infrastructure Design (deployment architecture)
Code Generation (implementation)
Starting with Unit 1: Shared Components - the foundation that all other units depend on. This unit provides file storage, logging, error handling, and validation components.

Ready to proceed with Functional Design for Unit 1?
```

raj-jain-aws reviewed (changes requested):
I tested this PR and am not seeing the multi-user steps kick in. One reason could be that this PR does not seem to be baselined from the latest commits. Also, it will be great if the PR can be baselined based on latest commits and only include the files that'll be actually merged.

SiddhJog reviewed (changes requested):
I will raise an issue that we will use as a guiding principle for this work.
We need to adhere to the current rule mechanism and not introduce new concepts (AUDIT etc) as a part of this.
Also for this feature PR lets just work out of a branch and not use the fork mechanism.
The code needs to be rebased before raising a PR.
Please have a look at file specific comments @raj-jain-aws

jschwellach commented:
Thanks for raising this.
I think there was a merging issue due to the move from internal code repo
to GitHub which caused all files from previous commits to be included.
I agree with the AGENTS.md file, that was a direct requirement during a
customer session.
The audit file split is necessary though for this to work. I'll clean up
the PR before X-Mas.

On Wed, 17 Dec 2025, 20:04 Siddhesh, ***@***.***> wrote:

> ***@***.**** requested changes on this pull request.
>
> I will raise an issue that we will use as a guiding principle for this
> work.
> We need to adhere to the current rule mechanism and not introduce new
> concepts (AUDIT etc) as a part of this.
> Also for this feature PR lets just work out of a branch and not use the
> fork mechanism.
> The code needs to be rebased before raising a PR.
> Please have a look at file specific comments @raj-jain-aws
> <https://github.com/raj-jain-aws>
> ------------------------------
>
> On AGENTS.md.example
> <https://github.com/awslabs/aidlc-workflows/pull/25#discussion_r2626796111>
> :
>
> I will raise an issue that we will use as a guiding principle for this
> work.
> We need to adhere to the current rule mechanism and not introduce new
> concepts (AUDIT etc) as a part of this.
> Also for this feature PR lets just work out of a branch and not use the
> fork mechanism.
> The code needs to be rebased before raising a PR.
> Please have a look at file specific comments @raj-jain-aws
> <https://github.com/raj-jain-aws>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/awslabs/aidlc-workflows/pull/25#pullrequestreview-3587412716>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AANFNIDVYGELHTEUMJYIPMT4CFBDJAVCNFSM6AAAAACPERJPVCVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTKOBXGQYTENZRGY>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

jschwellach committed:
feat: Add multi-user parallel development support to AI-DLC

This commit introduces comprehensive multi-user collaboration capabilities
integrated into the core AI-DLC workflow:

New Features:
- Multi-user parallel development with automatic branch management
- User-scoped audit files to prevent conflicts
- Unit assignment tracking and dependency-aware consolidation
- Consolidated project views with automatic indexing

New Files:
- AGENTS.md.example: Configuration template for multi-user behavior
- QUICK-REFERENCE.md: Quick commands and workflows reference
- multi-user-commands.md: Command handling rules
- multi-user-parallel.md: Parallel workflow orchestration rules

Updated Files:
- README.md: Added multi-user documentation and configuration sections
- units-generation.md: Added Step 20 for multi-user setup workflow
- Rule files enhanced with multi-user workflow integration
- All documentation uses current 'kiro-cli chat' command syntax

Multi-user functionality is built into the core workflow and automatically
available. After inception, users can choose solo work or enable team
collaboration with automatic unit assignment and consolidation.

jschwellach reviewed (commented):
I cleaned up the feature and squashed the commits
